src/category_theory/sites/sheaf.lean

```diff
@@ -0,0 +1,269 @@
+      ext x,
+      apply presieve.is_sheaf_for.valid_glue },
+    { intros y hy,
+      ext x,
+      apply (hP S hS).is_separated_for.ext,
+      intros Y' f hf,
+      rw [presieve.is_sheaf_for.valid_glue _ _ _ hf, ← hy _ hf],
+      refl } }
+end
+
+/--
+The category of sheaves taking values in Type is the same as the category of set-valued sheaves.
+-/
+@[simps]
+def Sheaf_equiv_SheafOfTypes : Sheaf J (Type v) ≌ SheafOfTypes J :=
+{ functor :=
+  { obj := λ S, ⟨S.1, (is_sheaf_iff_is_sheaf_of_type _ _).1 S.2⟩,
+    map := λ S₁ S₂ f, f },
+  inverse :=
+  { obj := λ S, ⟨S.1, (is_sheaf_iff_is_sheaf_of_type _ _).2 S.2⟩,
+    map := λ S₁ S₂ f, f },
+  unit_iso := nat_iso.of_components (λ X, ⟨𝟙 _, 𝟙 _, by tidy, by tidy⟩) (by tidy),
+  counit_iso := nat_iso.of_components (λ X, ⟨𝟙 _, 𝟙 _, by tidy, by tidy⟩) (by tidy) }
+
+instance : inhabited (Sheaf (⊥ : grothendieck_topology C) (Type v)) :=
+⟨(Sheaf_equiv_SheafOfTypes _).inverse.obj (default _)⟩
+
+end category_theory
+
+namespace category_theory
+
+open opposite category_theory category limits sieve classical
+
+namespace presheaf
+
+-- Under here is the equalizer story, which is equivalent if A has products (and doesn't
+-- make sense otherwise). It's described in https://stacks.math.columbia.edu/tag/00VL,
+-- between 00VQ and 00VR.
+
+variables {C : Type v} [small_category C]
+variables {A : Type u} [category.{v} A]
+variables (J : grothendieck_topology C)
+variables {U : C} (R : presieve U)
+variables (P : Cᵒᵖ ⥤ A)
+
+section
+
+variables [has_products A]
+
+def first_obj : A :=
+∏ (λ (f : Σ V, {f : V ⟶ U // R f}), P.obj (op f.1))
+
+variables [has_pullbacks C]
+
+/--
+The rightmost object of the fork diagram of https://stacks.math.columbia.edu/tag/00VM, which
+contains the data used to check a family of elements for a presieve is compatible.
+-/
+def second_obj : A :=
+∏ (λ (fg : (Σ V, {f : V ⟶ U // R f}) × (Σ W, {g : W ⟶ U // R g})),
+  P.obj (op (pullback fg.1.2.1 fg.2.2.1)))
+
+/-- The map `pr₀*` of https://stacks.math.columbia.edu/tag/00VM. -/
+def first_map : first_obj R P ⟶ second_obj R P :=
+pi.lift (λ fg, pi.π _ _ ≫ P.map pullback.fst.op)
+
+/-- The map `pr₁*` of https://stacks.math.columbia.edu/tag/00VM. -/
+def second_map : first_obj R P ⟶ second_obj R P :=
+pi.lift (λ fg, pi.π _ _ ≫ P.map pullback.snd.op)
+
+/--
+The left morphism of the fork diagram given in Equation (3) of [MM92], as well as the fork diagram
+of https://stacks.math.columbia.edu/tag/00VM.
+-/
+def fork_map : P.obj (op U) ⟶ first_obj R P :=
+pi.lift (λ f, P.map f.2.1.op)
+
+lemma w : fork_map R P ≫ first_map R P = fork_map R P ≫ second_map R P :=
+begin
+  apply limit.hom_ext,
+  rintro ⟨⟨Y, f, hf⟩, ⟨Z, g, hg⟩⟩,
+  simp only [first_map, second_map, fork_map, limit.lift_π, limit.lift_π_assoc, assoc,
+    fan.mk_π_app, subtype.coe_mk, subtype.val_eq_coe],
+  rw [← P.map_comp, ← op_comp, pullback.condition],
+  simp,
+end
+
+def is_sheaf' (P : Cᵒᵖ ⥤ A) : Prop := ∀ (U : C) (R : presieve U) (hR : generate R ∈ J U),
```

Comment:
In the sheaf on spaces library, sheaf conditions consistently live in `Type`, not `Prop`. Making it a `Type` would alleviate you from having to write `nonempty` here I think. Not sure which is better, but this might be a compatibility issue when we try to unify with sheaves on spaces, so we should decide at some point...

Reply:
This is a reasonable point - it was discussed a bit in the sheaves of types PR as well, and in the docstring there (https://github.com/leanprover-community/mathlib/blob/master/src/category_theory/sites/sheaf_of_types.lean#L73) - the summary is that it's a subsingleton condition and so it essentially has no benefit to being in Type as opposed to Prop, and it can make some universe troubles a little easier, which is more of a concern with sites.
I agree we should be consistent though - maybe @semorrison and @kbuzzard have thoughts?

Reply:
Since we started work on sheaves, I've become generally more enthusiastic about `Prop` and `nonempty`, and no longer see much point in staying in `Type` in this situation.

So yes, consistency would be great, but I don't think it's this PRs job to go change the old code, as long as we have a plan/intent to eventually unify these two developments which would achieve consistency ... later!

Reply:
Mathematically `is_sheaf` sounds like a Prop to me.

Reply:
I think the conclusion here is that we'll eventually change the notion in the sheaf on spaces library to `Prop`, so this should stay as it is: I'll resolve this conversation.